google-protobuf won't build on Linux after adding MinGW platform

### Discussed in https://github.com/rubygems/rubygems/discussions/7315

<div type='discussions-op-text'>

<sup>Originally posted by **AnyOldName3** December 20, 2023</sup>
I had a working Jekyll project that built fine under both real Linux and WSL2, but for reasons not worth getting into, needed to switch from WSL2 to Windows. I ran `bundle lock --add-platform x64-mingw-ucrt01`, the results of which can be seen as this commit: https://github.com/AnyOldName3/osg-website-replacement/commit/a72328e722c0ab38b0914794e767443a14210fb6. The line I believe to be most important changes `google-protobuf (3.25.1-x86_64-linux)` to `google-protobuf (3.25.1)`, and looking at the contents of the two packages, it appears the former has prebuilt `.so`s whereas the latter is source-only, so the shared library needs building.

As intended, this made the project work with Windows Ruby, however, it also produced the following errors on Linux:
```
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /mnt/c/vsg/osg-website-replacement/vendor/bundler/ruby/3.0.0/gems/google-protobuf-3.25.1/ext/google/protobuf_c
/usr/bin/ruby3.0 -I /usr/lib/ruby/vendor_ruby -r ./siteconf20231220-407-i7czk0.rb extconf.rb
creating Makefile

current directory: /mnt/c/vsg/osg-website-replacement/vendor/bundler/ruby/3.0.0/gems/google-protobuf-3.25.1/ext/google/protobuf_c
make DESTDIR\= clean

current directory: /mnt/c/vsg/osg-website-replacement/vendor/bundler/ruby/3.0.0/gems/google-protobuf-3.25.1/ext/google/protobuf_c
make DESTDIR\=
compiling convert.c
convert.c: In function ‘Msgval_IsEqual’:
convert.c:295:58: warning: passing argument 4 of ‘shared_Msgval_IsEqual’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  295 |                                             type_info.def.msgdef, &status);
      |                                             ~~~~~~~~~~~~~^~~~~~~
In file included from convert.c:21:
shared_convert.h:19:60: note: expected ‘upb_MessageDef *’ but argument is of type ‘const upb_MessageDef *’
   19 |                            upb_CType type, upb_MessageDef* msgdef,
      |                                            ~~~~~~~~~~~~~~~~^~~~~~
convert.c:299:5: error: format not a string literal and no format arguments [-Werror=format-security]
  299 |     rb_raise(rb_eRuntimeError, upb_Status_ErrorMessage(&status));
      |     ^~~~~~~~
convert.c: In function ‘Msgval_GetHash’:
convert.c:308:41: warning: passing argument 3 of ‘shared_Msgval_GetHash’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  308 |       val, type_info.type, type_info.def.msgdef, seed, &status);
      |                            ~~~~~~~~~~~~~^~~~~~~
In file included from convert.c:21:
shared_convert.h:23:48: note: expected ‘upb_MessageDef *’ but argument is of type ‘const upb_MessageDef *’
   23 |                                upb_MessageDef* msgdef, uint64_t seed,
      |                                ~~~~~~~~~~~~~~~~^~~~~~
convert.c:312:5: error: format not a string literal and no format arguments [-Werror=format-security]
  312 |     rb_raise(rb_eRuntimeError, upb_Status_ErrorMessage(&status));
      |     ^~~~~~~~
cc1: some warnings being treated as errors
make: *** [Makefile:245: convert.o] Error 1

make failed, exit code 2
```
This log is from using WSL2, but it also happens in CI: https://github.com/AnyOldName3/osg-website-replacement/actions/runs/7266847247/job/19829238652. I believe that's Arch running in Docker (on an Ubuntu host) whereas WSL2 is running Ubuntu, so dependency packages from the OS should be different - they're at least using different versions of Ruby.

`-Werror=format-security` is coming from `rbconfig.rb` on the WSL2 system, and I imagine that's where it's coming from on the probably-Arch VM, too. Conceivably, other compiler flags requested by that file could be responsible for the other errors. I don't know whether this is a case of a broken gem (i.e. that these flags should work), a packaging error for Ruby itself (i.e. that these flags shouldn't have ended up in `rbconfig.rb`, or some other kind of problem. I also don't know how I'd go about working around this, e.g. by somehow forcing `Gemfile.lock` to use the prebuilt Linux variant of the package on Linux.</div>

--- a/bundler/lib/bundler/definition.rb
+++ b/bundler/lib/bundler/definition.rb
@@ -312,10 +312,6 @@ module Bundler
       end
     end
 
-    def should_complete_platforms?
-      !lockfile_exists? && generic_local_platform_is_ruby? && !Bundler.settings[:force_ruby_platform]
-    end
-
     def spec_git_paths
       sources.git_sources.map {|s| File.realpath(s.path) if File.exist?(s.path) }.compact
     end
@@ -516,6 +512,10 @@ module Bundler
     end
 
     private
+
+    def should_complete_platforms?
+      !lockfile_exists? && generic_local_platform_is_ruby? && !Bundler.settings[:force_ruby_platform]
+    end
 
     def lockfile_exists?
       lockfile && File.exist?(lockfile)

--- a/bundler/lib/bundler/definition.rb
+++ b/bundler/lib/bundler/definition.rb
@@ -602,6 +602,8 @@ module Bundler
       @resolved_bundler_version = result.find {|spec| spec.name == "bundler" }&.version
       @platforms = result.add_extra_platforms!(platforms) if should_add_extra_platforms?
 
+      result.complete_platforms!(platforms)
+
       SpecSet.new(result.for(dependencies, false, @platforms))
     end
 

--- a/bundler/lib/bundler/definition.rb
+++ b/bundler/lib/bundler/definition.rb
@@ -513,7 +513,7 @@ module Bundler
 
     private
 
-    def should_complete_platforms?
+    def should_add_extra_platforms?
       !lockfile_exists? && generic_local_platform_is_ruby? && !Bundler.settings[:force_ruby_platform]
     end
 
@@ -600,7 +600,7 @@ module Bundler
       result = SpecSet.new(resolver.start)
 
       @resolved_bundler_version = result.find {|spec| spec.name == "bundler" }&.version
-      @platforms = result.complete_platforms!(platforms) if should_complete_platforms?
+      @platforms = result.add_extra_platforms!(platforms) if should_add_extra_platforms?
 
       SpecSet.new(result.for(dependencies, false, @platforms))
     end

--- a/bundler/lib/bundler/spec_set.rb
+++ b/bundler/lib/bundler/spec_set.rb
@@ -71,6 +71,12 @@ module Bundler
       platforms
     end
 
+    def complete_platforms!(platforms)
+      platforms.each do |platform|
+        complete_platform(platform)
+      end
+    end
+
     def validate_deps(s)
       s.runtime_dependencies.each do |dep|
         next if dep.name == "bundler"

--- a/bundler/lib/bundler/spec_set.rb
+++ b/bundler/lib/bundler/spec_set.rb
@@ -52,7 +52,7 @@ module Bundler
       specs.uniq
     end
 
-    def complete_platforms!(platforms)
+    def add_extra_platforms!(platforms)
       return platforms.concat([Gem::Platform::RUBY]).uniq if @specs.empty?
 
       new_platforms = @specs.flat_map {|spec| spec.source.specs.search([spec.name, spec.version]).map(&:platform) }.uniq.select do |platform|

--- a/bundler/lib/bundler/spec_set.rb
+++ b/bundler/lib/bundler/spec_set.rb
@@ -86,8 +86,7 @@ module Bundler
       less_specific_platform = new_platforms.find {|platform| platform != Gem::Platform::RUBY && platform === Bundler.local_platform }
       platforms.delete(Bundler.local_platform) if less_specific_platform
 
-      @sorted = nil
-      @lookup = nil
+      reset!
 
       platforms
     end
@@ -110,14 +109,14 @@ module Bundler
 
     def []=(key, value)
       @specs << value
-      @lookup = nil
-      @sorted = nil
+
+      reset!
     end
 
     def delete(specs)
       specs.each {|spec| @specs.delete(spec) }
-      @lookup = nil
-      @sorted = nil
+
+      reset!
     end
 
     def sort!
@@ -175,8 +174,8 @@ module Bundler
 
     def delete_by_name(name)
       @specs.reject! {|spec| spec.name == name }
-      @lookup = nil
-      @sorted = nil
+
+      reset!
     end
 
     def what_required(spec)
@@ -211,6 +210,11 @@ module Bundler
     end
 
     private
+
+    def reset!
+      @sorted = nil
+      @lookup = nil
+    end
 
     def valid_dependencies?(s)
       validate_deps(s) == :valid

--- a/bundler/lib/bundler/spec_set.rb
+++ b/bundler/lib/bundler/spec_set.rb
@@ -68,8 +68,6 @@ module Bundler
       less_specific_platform = new_platforms.find {|platform| platform != Gem::Platform::RUBY && platform === Bundler.local_platform }
       platforms.delete(Bundler.local_platform) if less_specific_platform
 
-      reset!
-
       platforms
     end
 
@@ -209,14 +207,18 @@ module Bundler
         end
 
         if platform_spec
-          new_specs << LazySpecification.from_spec(platform_spec)
+          new_specs << LazySpecification.from_spec(platform_spec) unless specs.include?(platform_spec)
           true
         else
           false
         end
       end
 
-      @specs.concat(new_specs.uniq) if valid_platform
+      if valid_platform && new_specs.any?
+        @specs.concat(new_specs)
+
+        reset!
+      end
 
       valid_platform
     end

--- a/bundler/lib/bundler/spec_set.rb
+++ b/bundler/lib/bundler/spec_set.rb
@@ -55,29 +55,11 @@ module Bundler
     def add_extra_platforms!(platforms)
       return platforms.concat([Gem::Platform::RUBY]).uniq if @specs.empty?
 
-      new_platforms = @specs.flat_map {|spec| spec.source.specs.search([spec.name, spec.version]).map(&:platform) }.uniq.select do |platform|
+      new_platforms = all_platforms.select do |platform|
         next if platforms.include?(platform)
         next unless GemHelpers.generic(platform) == Gem::Platform::RUBY
 
-        new_specs = []
-
-        valid_platform = lookup.all? do |_, specs|
-          spec = specs.first
-          matching_specs = spec.source.specs.search([spec.name, spec.version])
-          platform_spec = GemHelpers.select_best_platform_match(matching_specs, platform).find do |s|
-            s.matches_current_metadata? && valid_dependencies?(s)
-          end
-
-          if platform_spec
-            new_specs << LazySpecification.from_spec(platform_spec)
-            true
-          else
-            false
-          end
-        end
-        next unless valid_platform
-
-        @specs.concat(new_specs.uniq)
+        complete_platform(platform)
       end
       return platforms if new_platforms.empty?
 
@@ -214,6 +196,33 @@ module Bundler
     def reset!
       @sorted = nil
       @lookup = nil
+    end
+
+    def complete_platform(platform)
+      new_specs = []
+
+      valid_platform = lookup.all? do |_, specs|
+        spec = specs.first
+        matching_specs = spec.source.specs.search([spec.name, spec.version])
+        platform_spec = GemHelpers.select_best_platform_match(matching_specs, platform).find do |s|
+          s.matches_current_metadata? && valid_dependencies?(s)
+        end
+
+        if platform_spec
+          new_specs << LazySpecification.from_spec(platform_spec)
+          true
+        else
+          false
+        end
+      end
+
+      @specs.concat(new_specs.uniq) if valid_platform
+
+      valid_platform
+    end
+
+    def all_platforms
+      @specs.flat_map {|spec| spec.source.specs.search([spec.name, spec.version]).map(&:platform) }.uniq
     end
 
     def valid_dependencies?(s)

--- a/bundler/spec/lock/lockfile_spec.rb
+++ b/bundler/spec/lock/lockfile_spec.rb
@@ -1195,6 +1195,46 @@ RSpec.describe "the lockfile format" do
     G
   end
 
+  it "adds compatible platform specific variants to the lockfile, even if resolution fallback to RUBY due to some other incompatible platform specific variant" do
+    simulate_platform "arm64-darwin-23" do
+      build_repo4 do
+        build_gem "google-protobuf", "3.25.1"
+        build_gem "google-protobuf", "3.25.1" do |s|
+          s.platform = "arm64-darwin-23"
+        end
+        build_gem "google-protobuf", "3.25.1" do |s|
+          s.platform = "x64-mingw-ucrt"
+          s.required_ruby_version = "> #{Gem.ruby_version}"
+        end
+      end
+
+      gemfile <<-G
+        source "#{file_uri_for(gem_repo4)}"
+        gem "google-protobuf"
+      G
+      bundle "lock --add-platform x64-mingw-ucrt"
+
+      expect(lockfile).to eq <<~L
+        GEM
+          remote: #{file_uri_for(gem_repo4)}/
+          specs:
+            google-protobuf (3.25.1)
+            google-protobuf (3.25.1-arm64-darwin-23)
+
+        PLATFORMS
+          arm64-darwin-23
+          ruby
+          x64-mingw-ucrt
+
+        DEPENDENCIES
+          google-protobuf
+
+        BUNDLED WITH
+           #{Bundler::VERSION}
+      L
+    end
+  end
+
   it "persists the spec's specific platform to the lockfile" do
     build_repo2 do
       build_gem "platform_specific", "1.0" do |s|

--- a/bundler/spec/support/rubygems_ext.rb
+++ b/bundler/spec/support/rubygems_ext.rb
@@ -86,7 +86,7 @@ module Spec
         puts success_message
         puts
       else
-        system("git status --porcelain")
+        system("git diff")
 
         puts
         puts error_message

--- a/tool/bundler/dev_gems.rb.lock
+++ b/tool/bundler/dev_gems.rb.lock
@@ -11,6 +11,8 @@ GEM
     nokogiri (1.15.5)
       mini_portile2 (~> 2.8.2)
       racc (~> 1.4)
+    nokogiri (1.15.5-java)
+      racc (~> 1.4)
     nronn (0.11.1)
       kramdown (~> 2.1)
       kramdown-parser-gfm (>= 1.0.1, < 1.2)
@@ -81,6 +83,7 @@ CHECKSUMS
   mini_portile2 (2.8.5) sha256=7a37db8ae758086c3c3ac3a59c036704d331e965d5e106635e4a42d6e66089ce
   mustache (1.1.1) sha256=90891fdd50b53919ca334c8c1031eada1215e78d226d5795e523d6123a2717d0
   nokogiri (1.15.5) sha256=22448ca35dbcbdcec60dbe25ccf452b685a5436c28f21b2fec2e20917aba9100
+  nokogiri (1.15.5-java) sha256=5f87e71aaeb4f7479b94698737a0aacea77836b4805c7433b655e9565bd56cfe
   nronn (0.11.1) sha256=60305c7a42dcaf6bdd33210993a7e38087520717561da101bea1df7197546369
   parallel (1.23.0) sha256=27154713ad6ef32fa3dcb7788a721d6c07bca77e72443b4c6080a14145288c49
   parallel_tests (3.8.1) sha256=c781c0910be3b489411f24e3a3397d57f481d6ee4eb27dba2a1cd4b8a0967f06

